client/websocket/websocket.c

```diff
@@ -0,0 +1,696 @@
+#ifndef NDEBUG
+#define ws_debug(...) xdag_log(WS_LOG_FILE, XDAG_DEBUG   , __VA_ARGS__)
+#else
+#define ws_debug(...)
+#endif
+
+#define WS_DEFAULT_PORT 9999
+#define WS_DEFAULT_MAX_LIMIT 100
+
+static pthread_mutex_t g_mutex_sessions = PTHREAD_MUTEX_INITIALIZER;
+static pthread_mutex_t g_mutex_messages = PTHREAD_MUTEX_INITIALIZER;
+int g_websocket_running = 0;
+int g_websocket_port = WS_DEFAULT_PORT;
+int g_websocket_maxlimit = WS_DEFAULT_MAX_LIMIT;
+
+typedef struct {
+	int fd;
+	struct in_addr addr;
+	in_port_t port;
+	struct pollfd pfd;
+	wslay_event_context_ptr ctx;
+} ws_session;
+
+typedef struct ws_session_list_element_t {
+	ws_session *session;
+	struct ws_session_list_element_t *next;
+} ws_session_list_element;
+static ws_session_list_element *g_session_list = NULL;
+
+typedef struct ws_message_list_element_t {
+	char *message;
+	struct ws_message_list_element_t *next;
+} ws_message_list_element;
+static ws_message_list_element *g_message_list = NULL;
+
+// functions
+void ws_cleanup(void);
+void cleanup_threads(void *args);
+int ws_session_init(ws_session *session);
+int ws_session_communicate(ws_session *session);
+void *ws_handler_thread(void *args);
+void *ws_server_thread(void *args);
+
+void xdag_ws_message_append(char *message)
+{
+	if (!g_websocket_running) {
+		return;
+	}
+	ws_message_list_element *msg = malloc(sizeof(ws_message_list_element));
+	memset(msg, 0, sizeof(ws_message_list_element));
+	msg->message = strdup(message);
+	pthread_mutex_lock(&g_mutex_messages);
+	LL_APPEND(g_message_list, msg);
+	pthread_mutex_unlock(&g_mutex_messages);
+}
+
+static ws_message_list_element* ws_first_message(void)
+{
+	ws_message_list_element *first = NULL;
+	first = g_message_list;
+	pthread_mutex_lock(&g_mutex_messages);
+	if (first) {
+		LL_DELETE(g_message_list, first);
+	}
+	pthread_mutex_unlock(&g_mutex_messages);
+	return first;
+}
+
+static void ws_session_close(ws_session *session)
+{
+	ws_info("session from %s:%d closed\n", inet_ntoa(session->addr), ntohs(session->port));
+	shutdown(session->fd, SHUT_RDWR);
+	close(session->fd);
+	free(session);
+}
+
+static void ws_session_append(ws_session *session)
+{
+	assert(session);
+
+	ws_session_list_element *elem = NULL;
+	int count = 0;
+
+	pthread_mutex_lock(&g_mutex_sessions);
+	LL_COUNT(g_session_list, elem, count);
+	pthread_mutex_unlock(&g_mutex_sessions);
+
+	if (count >= g_websocket_maxlimit) {
+		ws_session_close(session);
+	} else {
+		elem = calloc(1, sizeof(ws_session_list_element));
+		if (elem == NULL) {
+			ws_err("[ws_session_append] alloca failed!");
+			ws_session_close(session);
+		} else {
+			elem->session = session;
+			ws_session_init(session); // init wslay websocket
+			pthread_mutex_lock(&g_mutex_sessions);
+			LL_APPEND(g_session_list, elem);
+			pthread_mutex_unlock(&g_mutex_sessions);
+			ws_info("%s connected\n", inet_ntoa(session->addr));
+		}
+	}
+}
+
+static void sha1(uint8_t *dst, const uint8_t *src, size_t src_length)
+{
+	SHA_CTX ctx;
```

Comment:
Do we really need SHA-1? This algorithm is a bit outdated.

Reply:
websocket protocol uses sha-1

Reply:
It's needed as described in RFC6455. Just old one.